### ISSUE TITLE
Use a more resilient mechanism for showing the icons on the Icons page

### ISSUE THREAD

**Issue by DanTup** [flutter/flutter#11073]
The source code for `icons.dart` looks like this:

```dart
  /// <p><i class="material-icons md-36">4k</i> &#x2014; material icon named "4k".</p>
  static const IconData four_k = const IconData(0xe072, fontFamily: 'MaterialIcons');

  /// <p><i class="material-icons md-36">ac_unit</i> &#x2014; material icon named "ac unit".</p>
  static const IconData ac_unit = const IconData(0xeb3b, fontFamily: 'MaterialIcons');

  /// <p><i class="material-icons md-36">access_alarm</i> &#x2014; material icon named "access alarm".</p>
  static const IconData access_alarm = const IconData(0xe190, fontFamily: 'MaterialIcons');
```

This HTML shows up in the IDE:

![Icons](https://user-images.githubusercontent.com/1078012/27770228-80f92dec-5f32-11e7-81b8-c5e018c54b1e.png)

I don't know what's expected here - even if I was able to get the IDE tooltips to handle the HTML (though [the docs say](https://www.dartlang.org/guides/language/effective-dart/documentation#avoid-using-html-for-formatting) avoid HTML) the icons wouldn't render correctly because the CSS isn't being loaded in the tooltips.

**Comment by Hixie** [comment]
This seems like an IDE problem? :-)

@DanTup I'm going to close this since I'm not sure what you want us to change, but please don't hesitate to reach out to me if you want to discuss his further. We should definitely find a way to make it work in all our IDEs!

**Comment by DanTup** [comment]
TL;DR: I still thing this is valid and should be addressed.

My suggestion is that we should not have HTML inside the comments in the source code where possible. [The docs](https://www.dartlang.org/guides/language/effective-dart/documentation#avoid-using-html-for-formatting) say to avoid it (though avoid is a wooly term). In this case, even if the IDE could render HTML, these HTML chunks required a specific stylesheet/font to be loaded to work, which is even less likely to be available to an IDE in its tooltips.

What would be better is having *MarkDown* image tags, like the colours code does:

<img width="727" alt="screen shot 2018-05-02 at 8 18 11 am" src="https://user-images.githubusercontent.com/1078012/39510355-632a8592-4de1-11e8-8962-58a0e635d8f0.png">

This has no requirement for CSS/fonts and should work anywhere that is rendering Markdown correctly. The images are already available online (though these are transparent so might not play nice in light/dark themes - maybe we should have copies; it'd also remove reliance on them staying at these urls?):

https://storage.googleapis.com/material-icons/external-assets/v4/icons/svg/ic_ac_unit_white_36px.svg
https://storage.googleapis.com/material-icons/external-assets/v4/icons/svg/ic_ac_unit_black_36px.svg

For now, I'm working around this in Dart Code by regexing them out and replacing with a markdown image tag; but it could break if this HTML changes.

FWIW, IntelliJ/AS don't render these properly either - they strip the HTML but leave the text, so you just get this weird label/alt text:

![intellij icon](https://user-images.githubusercontent.com/1078012/37088348-f3a99c48-21f4-11e8-940d-b7ae254601e0.png)

**Comment by Hixie** [comment]
Using images is a non-starter, because of this page:
https://master-docs-flutter-io.firebaseapp.com/flutter/material/Icons-class.html

Using the font is what we want, but I agree that the current mechanism isn't good.

@jcollins-g do you have any suggestions for how we could mark up these cases so that they work both in the HTML output and in IDEs? It's ok if it requires work on the IDE side, but it needs to be something more resilient than relying on a magic HTML sequence that happens to trigger some CSS.

cc @devoncarew

**Comment by jcollins-g** [comment]
@Hixie Why are images are a non-starter here?  The example @DanTup gave seems not particularly worse than the situation in class Icons from a coding perspective, and it would definitely render similarly across.

```dart
  /// <p><i class="material-icons md-36">360</i> &#x2014; material icon named "360".</p>
  static const IconData threesixty = const IconData(0xe577, fontFamily: 'MaterialIcons');

  /// <p><i class="material-icons md-36">3d_rotation</i> &#x2014; material icon named "3d rotation".</p>
  static const IconData threed_rotation = const IconData(0xe84d, fontFamily: 'MaterialIcons');
```

However, assuming they _are_ a nonstarter for reasons @Hixie is about to explain to me, there's always the potential for more magical Dartdoc-only strings, as scary as that can be sometimes.  Maybe something like this:

```dart
  /// {@style material-icons 3d_rotation} -- material icon named "3d rotation".
  static const IconData threed_rotation = const IconData(0xe84d, fontFamily: 'MaterialIcons');
```

The challenge for dartdoc here is to do this in a way that's not accumulating large amounts of hardcoding.  Probably would want to have a configurable mapping of style name to Markdown/HTML wrapping code, e.g.:

```dart
{ 'material-icons':  '<i class="material-icons md-36">%s%</i>' }
```
or some such thing.  A few of these mappings would be supported by Dartdoc's default style; others would need to be added by users of the style sheet.  Dartdoc's new configuration system at least will make this less ugly than it otherwise would be.

However, I have no idea how hard it would be to implement this in an IDE -- it might be easier than it would be to do magical CSS inference though.  It's possible we could have style snippets defined for each of the supported **`@style`** tags in dartdoc's config, and maybe the IDE could read it.

As an aside, if there are too many more of these magical dartdoc directives added at some point I'll want to turn this into a real parser/language but we haven't quite hit critical mass there yet.  In general, the parsing and interpretation of dartdoc-specific markdown hacks^H^H^H^H^Hextensions probably needs dartdoc integrated more strongly with the analyzer (e.g. a dartdoc analyzer plugin).   So anything we do here is going to be at least a little ugly.

**Comment by DanTup** [comment]
> Why are images are a non-starter here?

I was going to ask the same; they're already used for the colours? It seems like the least effort to implement on both sides to me (neither the website nor the IDE would need to do anything).

> `/// {@style material-icons 3d_rotation} -- material icon named "3d rotation".`

I'm not sure I understand what you're proposing, but we can't rely on any CSS classes or fonts in VS Code (and I expect that applies to all IDEs - IntelliJ isn't HTML-based so probably also cannot). So we'd still be left having to regex things out; which is only slightly better than what we have today (in that it might be slightly less fragile).

**Comment by DanTup** [comment]
Gotcha. I still personally think (markdown) images would be perfect; the docs are already markdown so whoever is consuming them is already handling that. Markdown image tags (as used in colours) just work for free without anyone needing to do custom parsing of the dartdocs.

**Comment by jcollins-g** [comment]
Markdown images make a lot of sense to me too.  I'm interested to hear why @Hixie thinks they are not viable.

**Comment by devoncarew** [comment]
For IntelliJ, we special case knowledge of Flutter colors and Icons, and have generated material icons in the flutter-intellij repo (we rely more on analyzing the code then annotations in the comments).

**Comment by DanTup** [comment]
I'm thinking of doing something similar in Dart Code to get them in the gutter; but if we intend for that to be a replacement, we'd still need to regex out the HTML. IntelliJ just seems to strip the HTML tags but leave what's between them, so it looks a bit wonky (see screenshot further up), so that should probably be tidied up if we keep the HTML here.

**Comment by devoncarew** [comment]
But we're happy to update if there's a more structured way that's easier to share between tools.

**Comment by Hixie** [comment]
Images are a non-starter because there's 1,970 of them on that page.

We could use dartdoc templates and have a different file defining the templates for HTML+CC-aware dartdoc processors and those that can't do HTML.

**Comment by DanTup** [comment]
> Images are a non-starter because there's 1,970 of them on that page.

Ah, I see what you mean :-)

> We could use dartdoc templates

This sounds interesting; I didn't know about these. Would they need to be resolved by the clients, or handled in the analysis server? If we'd need to parse and understand them in the editor (which might not be written in Dart, so can't just use existing libraries) it might end up being just as complicated. That said, if these templates may already be in use, maybe we need to do this anyway and currently they're being handled incorrectly?

Another option could be the opposite of what's there today - use markdown images, and then have the website regex them out for the material design font (this way it's all pure markdown for consumers, but those that want additional optimisations can have them; and there's also a benefit of the markdown regex being less fragile than the HTML one).

Btw, I don't think this is high priority - the solution today kinda works (except for us always using a white icon in VS Code, so people with a light theme won't see them so well). I only suggested re-opening because I think the current solution puts an unnecessary (but fairly minor) burden on editor integrations. I think we can do better, but there are definitely bigger fish to fry - there are very few people integrating into editors compared to end users.

**Comment by markusaksli-nc** [comment]
Hi @Hixie 
This issue no longer seems relevant in Intellij IDEA 2020.2, Android Studio 4.0.1, and VS Code 1.47.
<img src="https://user-images.githubusercontent.com/68219924/87780668-e04b2900-c837-11ea-91d9-a7f2353f93ec.png" height="300" />
<img src="https://user-images.githubusercontent.com/68219924/87780674-e17c5600-c837-11ea-81ed-6b5a26d5934f.png" height="300" />
<img src="https://user-images.githubusercontent.com/68219924/87780673-e0e3bf80-c837-11ea-9fe8-60f287abf89a.png" height="300" />
Should it still remain open?
Thank you

**Comment by DanTup** [comment]
Currently VS Code is handling this by using regular expressions to extract the HTML code:

```ts
const materialIconRegex = new RegExp(
	`(?:${escapeRegExp("<p>")})?`
	+ escapeRegExp('<i class="material-icons md-36">')
	+ "([\\w\\s_]+)"
	+ escapeRegExp('</i> &#x2014; material icon named "')
	+ "([\\w\\s_]+)"
	+ escapeRegExp('".')
	+ `(?:${escapeRegExp("</p>")})?`,
	"gi",
);
```

This works, but isn't ideal. If that HTML changes in the source (which it has in the past), the VS Code previews will break until the regex is updated (and released) - the regex may also need to support multiple SDK versions at the same time.

There are other editors rendering this that probably don't do the same regex - actually, this reminds me that the LSP server doesn't, so this will regress in VS Code if we don't fix it there (which will involve putting this regex in another place).

I don't think it's a big problem, but it would still be nice to have something better.

**Comment by guidezpl** [comment]
Working on https://github.com/flutter/flutter/issues/102560 at the moment, which will be in a 3P package, and HTML + CSS really seems like the ideal solution. A couple CSS files is all you need to display all icons. 

I know there are some limitations for VS Code to work around, but still think we can avoid some pitfalls like having to manage 6000 (yes, up from 1970) .pngs, or committing any static files at all. 

In any case, I think this issue can be closed, since work relating to IDEs at https://github.com/dart-lang/sdk/issues/47667.
